webapp/src/ts/services/training-cards.service.ts

```diff
@@ -0,0 +1,124 @@
+import { SessionService } from '@mm-services/session.service';
+import { RouteSnapshotService } from '@mm-services/route-snapshot.service';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class TrainingCardsService {
+  private globalActions;
+  private readonly TRAINING_PREFIX = 'training:'
+
+  constructor(
+    private store: Store,
+    private xmlFormsService: XmlFormsService,
+    private dbService: DbService,
+    private modalService: ModalService,
+    private sessionService: SessionService,
+    private routeSnapshotService: RouteSnapshotService,
+  ) {
+    this.globalActions = new GlobalActions(store);
+  }
+
+  private canOpenTraining() {
+    const routeSnapshot = this.routeSnapshotService.get();
+    return !this.sessionService.isDbAdmin() &&
+      (!document.getElementsByClassName('enketo').length ||
+        (routeSnapshot?.data?.tab === 'tasks' && !routeSnapshot?.params?.id));
+  }
+
+  private getAvailableTrainingForms(xForms, userCtx) {
+    const today = new Date();
+
+    return xForms
+      .map(xForm => ({
+        id: xForm._id,
+        code: xForm.internalId,
+        startDate: xForm.start_date,
+        duration: xForm.duration,
+        userRoles: xForm.user_roles,
+      }))
+      .filter(form => {
+        const hasRole = form.userRoles?.find(role => userCtx.roles.includes(role));
+        if (form.userRoles?.length && !hasRole) {
+          return false;
+        }
+
+        const startDate = new Date(form.startDate);
+        if (!form.duration) {
+          return startDate < today;
+        }
+
+        const endDate = new Date(form.startDate);
+        endDate.setDate(endDate.getDate() + form.duration);
+        return startDate < today && endDate > today;
+      })
+      .sort((a, b) => a.startDate - b.startDate);
+  }
+
+  private async getCompletedTrainings(userCtx) {
+    const docs = await this.dbService
+      .get()
+      .allDocs({
+        include_docs: true,
+        startkey: `${this.TRAINING_PREFIX}${userCtx.name}:`,
+        endkey: `${this.TRAINING_PREFIX}${userCtx.name}:\ufff0`,
+      });
+
+    if (!docs?.rows?.length) {
+      return;
+    }
+
+    return new Set(docs.rows.map(row => row?.doc?.form));
+  }
+
+  private async handleTrainingCards(error, xForms) {
+    if (error) {
+      console.error('Error fetching training cards.', error);
+      return;
+    }
+
+    try {
+      const userCtx = this.sessionService.userCtx();
+      let trainingForms = this.getAvailableTrainingForms(xForms, userCtx);
+      if (!trainingForms?.length) {
+        return;
+      }
+
+      const completedTrainings = await this.getCompletedTrainings(userCtx);
+      if (completedTrainings) {
+        trainingForms = trainingForms.filter(form => !completedTrainings.has(form.code));
+      }
+
+      if (!trainingForms.length) {
+        return;
+      }
+
+      this.globalActions.setTrainingCard(trainingForms[0].code);
+      this.modalService.show(TrainingCardsComponent, { backdrop: 'static' });
```

Comment:
If I'm understanding this correctly, then if a new card is found then it'll immediately pop up on top of everything, without warning. Is that right? Is that what we want?

Reply:
My understanding of this code is it won't show up when the page loads, but will show up when the training form replicates down. In other words they will start the app, 10 seconds later the form will download, by then they're halfway through completing a task, and THEN we pop up and tell them to do the training.

Is that what the code is doing now? You can easily test this by logging in, uploading a training form, doing a sync, and see what happens.

I think a better user experience would be to show the training the next time the app is reloaded. For one thing at least they're not in the middle of a task and we interrupt them. But secondly, if there are app changes (like the icons or fabs) then the training will show _just_ before they see them which is perfect timing.

Basically I don't think you need to subscribe to `xmlFormsService` at all - just check for new forms at startup.

Reply:
Once the training is replicated in the user's device, if they close the app and open again or just reload, the training is displayed. As well if it just replicated and it needs to show immediately because of how it was configured. 

>I think a better user experience would be to show the training the next time the app is reloaded. For one thing at least they're not in the middle of a task and we interrupt them. But secondly, if there are app changes (like the icons or fabs) then the training will show just before they see them which is perfect timing.

Unless the partner wants to train in workflow changes, that doesn't trigger the app update message for reloading, and let's say they want to train immediately as soon as the workflow changes are synced. Not sure if we want to cover this case now or just label it as edge case and just display it next time the app loads.
I'm going to link the team to this question for confirmation and get back with feedback.

Reply:
I took the topic to the team ([care team channel](https://medic.slack.com/archives/C02GME3BQE6/p1654668461417949)), the team believes that the scenario below: 

> Partner wants to train in workflow changes, that doesn't trigger the app update message for reloading. They want to train immediately as soon as the updated workflow is synced into the user's device

.. might happen more frequently than "training only in app load". The team would like to keep both and gather user feedback to further iterate based on user behaviour.